sbt

```diff
@@ -534,11 +577,19 @@ fi
 # traceLevel is 0.12+
 [[ -n "$trace_level" ]] && setTraceLevel
 
+# options before a -- may be interpreted as options for itself by the
```

Comment:
The "for itself by the coursier-based launcher" is confusing.

And I kind of don't understand how this works.  What is `-Dcoursier.sbt-launcher.parse-args=true` and why is it needed and not the default?

If I understand this properly it means:
* if I specify `SBTX_COURSIER=true` (which sets `coursier_launcher_version` to true)
* and also specify a coursier argument, let's say `-CshortCircuitSbtMain=true`
* then the script will run:

```bash
java \
  -Dcoursier.sbt-launcher.parse-args=true \
  -jar ~/.sbt/launchers/coursier_1.2.4/sbt-launch.jar \
  -CshortCircuitSbtMain=true \
  --
```

Is that right?

Reply:
Just changed that to
```bash
[[ -z "$coursier_launcher_version" ]] || {
  addJava "-Dcoursier.sbt-launcher.parse-args=true"
  addCoursier "--"
}
```
which should be a bit clearer.

The `coursier.sbt-launcher.parse-args` Java property tells the launcher to parse itself arguments before a `--` (we pass the `-C…` args to it this way).

Reply:
Ok, and why is it opt-in instead of the default behaviour?

And why is the `--` required?

Couldn't csbt just always handle `-C` options (removing them from being passed to `xMain`) and handle a user-input `--`?  I'm thinking perhaps `sbt-extras` shouldn't add the `--` itself.

Reply:
That's a good point. I guess the launcher could handle the `-C` options itself, yes. I'm going to look into that.

Initially, `coursier.sbt-launcher.parse-args` was added in the context of https://github.com/coursier/sbt-launcher/issues/26 (so that it doesn't try to parse arguments if not asked to do so, and IntelliJ calling it with `--addSbtPlugin …` works fine).

Reply:
I think the launcher should handle arg parsing like sbt-extras does: using the equivalent of a `residual_args` that it forwards on.  So when it sees a `--addPluginSbtFile` it doesn't understand it saves it for forwarding, then moves to the next argument (the argument to `--addPluginSbtFile`) which it doesn't understand and saves that one too.

As a safeguard, it could handle a `--` as an instruction to pass the remaining arguments upstream (it doesn't pass the `--` upstream).  (Side-thought: maybe it should also handle `-C--` also/instead? 😕)